### PR TITLE
[WS-Passive-Scan] Add Error Application Disclosure scan 

### PR DESCRIPTION
**Pull request by ManosMagnus**
This Script analyzes incoming websocket messages for error messages.

____________________________________________________
TODOs:

- [x]  Add help content
- [x] Based on  #2089

### CODE REVIEW

**Comment by ManosMagnus** [comment]
Ready for review

File: addOns/websocket/src/main/javahelp/org/zaproxy/zap/extension/websocket/resources/help/contents/pscanrules.html [hunk 1]
```diff
@@ -11,5 +11,26 @@ <H1>WebSocket Passive Scan Rules</H1>
...
+            <tr><td>System.Data.OleDb.OleDbException: Syntax error in string in query expression 'User ID = ? And Password = ?\'</td><td>True Positive</td></tr>
+            <caption>Examples</caption>
+        </table>
+        <br>
+        <table border="1"  width = "500">
+            <caption>Default Values</caption>
+            <tr><th>Risk</th><td>Medium</td></tr>
+            <tr><th>Confidence</th><td>Medium</td></tr>
+            <tr><th>CWE ID</th><td>209: Information Exposure Through an Error Message</td></tr>
+            <tr><th>WASC ID</th><td>200: Information Leakage</td></tr>
```
Comment (kingthorin) [inline addOns/websocket/src/main/javahelp/org/zaproxy/zap/extension/websocket/resources/help/contents/pscanrules.html]: 13?

File: addOns/websocket/src/main/zapHomeFiles/scripts/templates/websocketpassive/application_error_scanner_regex.js [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+ to launch further attacks against the web application. The alert\
+ could be a false positive if the error message is found inside a\
+ documentation page.")
+                .setSolution("Review the source code of this page. Implement custom \
+ error pages. Consider implementing a mechanism to provide a unique\
+ error reference/identifier to the client (browser) while logging the\
+ details on the server side and not exposing them to the user.")
+                .setEvidence(String(matcher.group()))
+                .setCweId(209) // Information Exposure Through an Error Message
+                .setWascId(200) //Information Leakage
```
Comment (kingthorin) [inline addOns/websocket/src/main/zapHomeFiles/scripts/templates/websocketpassive/application_error_scanner_regex.js]: 13?

File: addOns/websocket/src/main/zapHomeFiles/scripts/templates/websocketpassive/application_error_scanner.js [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+ to launch further attacks against the web application. The alert\
+ could be a false positive if the error message is found inside a\
+ documentation page.")
+                    .setSolution("Review the source code of this page. Implement custom \
+ error pages. Consider implementing a mechanism to provide a unique\
+ error reference/identifier to the client (browser) while logging the\
+ details on the server side and not exposing them to the user.")
+                    .setEvidence(evidence)
+                    .setCweId(209) // Information Exposure Through an Error Message
+                    .setWascId(200) // Information Leakage
```
Comment (kingthorin) [inline addOns/websocket/src/main/zapHomeFiles/scripts/templates/websocketpassive/application_error_scanner.js]: 13?

File: addOns/websocket/src/main/javahelp/org/zaproxy/zap/extension/websocket/resources/help/contents/pscanrules.html (hunk 1 above)
Comment (ManosMagnus) [reply addOns/websocket/src/main/javahelp/org/zaproxy/zap/extension/websocket/resources/help/contents/pscanrules.html]: Fixed in 10e54dbda

File: addOns/websocket/src/main/zapHomeFiles/scripts/templates/websocketpassive/application_error_scanner.js (hunk 3 above)
Comment (ManosMagnus) [reply addOns/websocket/src/main/zapHomeFiles/scripts/templates/websocketpassive/application_error_scanner.js]: Fixed in 10e54dbda

File: addOns/websocket/src/main/zapHomeFiles/scripts/templates/websocketpassive/application_error_scanner_regex.js (hunk 2 above)
Comment (ManosMagnus) [reply addOns/websocket/src/main/zapHomeFiles/scripts/templates/websocketpassive/application_error_scanner_regex.js]: Fixed in 10e54dbda

File: addOns/websocket/src/main/zapHomeFiles/scripts/templates/websocketpassive/Application Error Scanner Regex.js [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+patterns.forEach(function(pattern){
+    errorPatterns.push(java.util.regex.Pattern.compile(pattern));
+});
+
+function scan(helper,msg) {
+
+    if(msg.opcode != OPCODE_TEXT || msg.isOutgoing){
+        return;
+    }
+    var message = String(msg.getReadablePayload()).valueOf();
```
Comment (thc202) [inline addOns/websocket/src/main/zapHomeFiles/scripts/templates/websocketpassive/Application Error Scanner Regex.js]: Remove `valueOf()` (same in the other script).

File: addOns/websocket/src/main/zapHomeFiles/scripts/templates/websocketpassive/Application Error Scanner Regex.js [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+    errorPatterns.forEach(function(pattern){
+        var matcher = pattern.matcher(message);
+        while(matcher.find()){
+            helper.newAlert()
+                .setRiskConfidence(RISK_MEDIUM, CONFIDENCE_MEDIUM)
+                .setName("Application Error Disclosure via WebSockets (regex-script)")
+                .setDescription("This page contains an error/warning message that \
```
Comment (thc202) [inline addOns/websocket/src/main/zapHomeFiles/scripts/templates/websocketpassive/Application Error Scanner Regex.js]: - This (and solution) needs to be tweaked to not say "page" (same for the other alert).
 - The space before the backslash should be removed (same in some of the following lines).

File: addOns/websocket/src/main/javahelp/org/zaproxy/zap/extension/websocket/resources/help/contents/pscanrules.html [hunk 6]
```diff
@@ -11,5 +11,26 @@ <H1>WebSocket Passive Scan Rules</H1>
         <h2 id="scripts">Scripts</h2>
         Scripts which are included by default in the add-on and they implement the following WebSocket passive scan rules:
 
+        <h3>Information Disclosure: Application Errors</h3>
+
+        This passive scanner checks incoming WebSocket message payload for known Application Error messages. Access to such details may provide a malicious individual with mean by which to further abuse the web site. They may also leak data not specifically meant for end user consumption.<br>
```
Comment (thc202) [inline addOns/websocket/src/main/javahelp/org/zaproxy/zap/extension/websocket/resources/help/contents/pscanrules.html]: `with means`?

File: addOns/websocket/src/main/javahelp/org/zaproxy/zap/extension/websocket/resources/help/contents/pscanrules.html [hunk 7]
```diff
@@ -11,5 +11,26 @@ <H1>WebSocket Passive Scan Rules</H1>
...
+
+        <br>
+        <table border="1"  width = "500">
+            <tr><th>Use case</th><th>Output</th></tr>
```
Comment (thc202) [inline addOns/websocket/src/main/javahelp/org/zaproxy/zap/extension/websocket/resources/help/contents/pscanrules.html]: Change to `Outcome`?

File: addOns/websocket/src/main/zapHomeFiles/scripts/templates/websocketpassive/Application Error Scanner Regex.js (hunk 4 above)
Comment (ManosMagnus) [reply addOns/websocket/src/main/zapHomeFiles/scripts/templates/websocketpassive/Application Error Scanner Regex.js]: Fixed in 396d1266c

File: addOns/websocket/src/main/zapHomeFiles/scripts/templates/websocketpassive/Application Error Scanner Regex.js (hunk 5 above)
Comment (ManosMagnus) [reply addOns/websocket/src/main/zapHomeFiles/scripts/templates/websocketpassive/Application Error Scanner Regex.js]: Fixed in 396d1266c

File: addOns/websocket/src/main/javahelp/org/zaproxy/zap/extension/websocket/resources/help/contents/pscanrules.html (hunk 6 above)
Comment (ManosMagnus) [reply addOns/websocket/src/main/javahelp/org/zaproxy/zap/extension/websocket/resources/help/contents/pscanrules.html]: Fixed in 11f31ed74

File: addOns/websocket/src/main/javahelp/org/zaproxy/zap/extension/websocket/resources/help/contents/pscanrules.html (hunk 7 above)
Comment (ManosMagnus) [reply addOns/websocket/src/main/javahelp/org/zaproxy/zap/extension/websocket/resources/help/contents/pscanrules.html]: Fixed in 11f31ed74

**Comment by ManosMagnus** [comment]
Rebased

**Comment by thc202** [comment]
Thanks!

**Comment by ManosMagnus** [comment]
Thank you! Commits are squashed

File: addOns/websocket/src/main/zapHomeFiles/scripts/templates/websocketpassive/Application Error Scanner Regex.js [hunk 8]
```diff
@@ -0,0 +1,69 @@
+// * This Script analyzes incoming websocket messages for error messages with a set of regular expressions
+
+// * Based on org.zaproxy.zap.extension.pscanrules.ApplicationErrorScanner
+// * Application error strings are equal to (characters '/' is escaped -> '//'):
```
Comment (kingthorin) [inline addOns/websocket/src/main/zapHomeFiles/scripts/templates/websocketpassive/Application Error Scanner Regex.js]: The patterns actually use the opposite slash `\` > `\\`

Comment (ManosMagnus) [reply addOns/websocket/src/main/zapHomeFiles/scripts/templates/websocketpassive/Application Error Scanner Regex.js]: Comment resolved

**Pull request merged**